### DISCUSSION TITLE
Where can I find documentation for LFRic apps software infrastructure?

### DISCUSSION THREAD

**Discussion by MichaelWhitall** [MetOffice/simulation-systems#251] [LFRic]
Hi everyone!

I'm currently attempting my first sci/tech review of a significant lfric_apps ticket, but rapidly concluding that on the sci/tech review form, in the box that says "Do you understand the area of the code and why it is being altered?", the only honest answer I can give for some of the changes is "no".  I'm finding changes within vast tracts of ".x90" files (whatever that means) which contain hardly any comments, and some python and other files which don't even contain a header outlining their method or purpose.  Clearly I really need to go away and read the documentation that explains the whole software infrastructure of LFRic before I can make sense of this.  Does anyone know where that is?

On here:
https://code.metoffice.gov.uk/trac/lfric_apps
There is a link to documentation here:
https://github.com/pages/auth?nonce=ae82a45b-806c-4e1d-9816-2e7f742d5b0d&page_id=41170540&path=Lw
But following this link I just get a "permission denied" error.

I'm also confused about the working practices with LFRic apps, as various bits of the well-trodden process in the UM just seem to be missing (e.g. there is no code-owner approvals table in the ticket summary template so I cannot see whether the affected code-owners have actually seen the change I'm reviewing and approve).  Any links to the documentation which clarifies this gratefully received :)

Cheers!
Mike

**Comment by jennyhickson** [comment]
Hi Mike,

I appreciate the structure and code in LFRic is overwhelming to get used to.  x90 files are those that will be processed by psyclone to become Fortan code - it'd be worth finding some docs that talk about algorithms and kernels which will help explain this. The lack of comments is frusturating and I'd fully support any request for more of those!

The documentation you linked to should give you good overviews of the structure of LFRic code. That documentation is still in the process of being developed which is why its internal only, but should be visible to you if you have a metoffice github account (you might need to be signed in for that to work). If not then there are older versions which still contain lots of good details available on the LFRic Core wiki pages here https://code.metoffice.gov.uk/trac/lfric/wiki/LFRicInfrastructure.

With respect to the working practices we're still bringing LFRic Apps up to speed with what you're used to in the UM. For example there is no formal sign off process for the code owners in LFRic Apps at the moment and theres a good chance they've not all been consulted. If you're looking at a large change then it'll be worth asking the dev to do that! If you have other working practices questions I'm happy to chat you through them. 

Hope that helps

**Reply by MichaelWhitall** [reply]
Thanks lots for the swift response :)  I'm confused as I am signed into github (hence I was able to post this discussion), but still can't access the lfric_apps documentation.  I'll have a look at the older page, many thanks for that...

Cheers!
Mike

**Reply by jennyhickson** [reply]
its not just github access, but being linked to the MetOffice github organisation which you might be lacking. 

A copy of the documentation has just been deployed here https://www-nwp/~lfric/lfric_core_docs/index.html which hopefully helps

**Comment by stevemullerworth** [comment]
As Jenny says, if you have a Met Office approved Github account then you should be able to access the documentation link which is in a Private repository (as opposed to this repo which is public). I assume a ServiceNow request is required to upgrade existing GitHub accounts. You would need to have two-factor authentication on your account to permit this.

I am hopeful that we will publish the documents externally soon.

Using Jenny's link, start with the User Documentation (a User is someone who develops or reviews LFRic applications). 

Broadly, the .x90 files are algorithms that hold the Fortran code that says how fields as a whole are treated. The work is done in kernels that are referenced in the "call invoke" calls in the algorithm. Simply put, the kernels replace the loops over the spatial domain that you would find in the UM. The code is converted to .f90 by PSyclone, which replaces the "call invoke" calls with calls to generated code that interfaces to the kernels.

Lack of comments or in code are something you would raise with the developer.
